### PR TITLE
Issue 397 Metric reporter to log debug messages if configured number …

### PR DESCRIPTION
**Pull request by manmagic3**
Metric inspector to read metrices every 5 seconds. It will check if number of failed repairs in a time window (default 30 minutes) exceed to that configured value in yaml file. If identifies that there are these many repairs in defined time window, then debug messages are printed with some of the metrices,showing status of the system at that moment. Also the timewindow is reconfigured to start next time window.

### CODE REVIEW

File: application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/config/metrics/StatisticsConfig.java [hunk 1]
```diff
@@ -25,6 +25,12 @@ public class StatisticsConfig
     private ReportingConfigs myReportingConfigs = new ReportingConfigs();
     private String myMetricsPrefix = "";
 
+    private int myRepairFailureCountForReporting = 0;
+
+    private int myTimeWindowSizeinMinsForReporting = 0;
```
Comment (jwaeab) [inline application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/config/metrics/StatisticsConfig.java:30]: Camel case; "...SizeInMins..."

File: core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/metrics/MetricInspector.java [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+public final class MetricInspector
+{
+
+    private final MeterRegistry myMeterRegistry;
+    private long myRepairFailureCountSinceLastReport = 0;
+
+    private static final int REPEAT_INTERVAL_PERIOD = 5000;
+
+     @VisibleForTesting
```
Comment (jwaeab) [inline core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/metrics/MetricInspector.java:34]: Indentations

File: core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/metrics/MetricInspector.java [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    }
+
+    private long myTotalRecordFailures = 0;
+
+    private final int myRepairFailureThreshold;
+
+    private final int myRepairFailureTimeWindow;
+
+
+    private LocalDateTime recordingStartTimestamp = LocalDateTime.now();
```
Comment (jwaeab) [inline core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/metrics/MetricInspector.java:59]: myRecordingStartTimestamp

File: core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/metrics/MetricInspector.java [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+            io.micrometer.core.instrument.Timer nodeRepairSessions = myMeterRegistry
+                     .find(TableRepairMetricsImpl.NODE_REPAIR_SESSIONS)
+                    .tags("successful", "false")
+                    .timer();
+            if (nodeRepairSessions != null)
+            {
+                myTotalRecordFailures = nodeRepairSessions.count();
+            }
+
+            if (myTotalRecordFailures - myRepairFailureCountSinceLastReport > myRepairFailureThreshold)
```
Comment (jwaeab) [inline core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/metrics/MetricInspector.java:99]: Will this eventually overflow?

File: core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/metrics/MetricInspector.java [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+            if (recordingStartTimestamp.isBefore(thirtyMinutesAgo))
+            {
+                myRepairFailureCountSinceLastReport = myTotalRecordFailures;
+                recordingStartTimestamp = LocalDateTime.now();
+            }
+
+
+        }
+
+
```
Comment (jwaeab) [inline core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/metrics/MetricInspector.java:134]: Lots of extra line breaks here.

File: core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/utils/StatusLogger.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+    static final String NODE_REPAIRED_RATIO = "node.repaired.ratio";
+
+
+/*
+Log the state of the table, if repairs for a table failed for more than time defined
+in configuration in ecc.yml
+//
+
+ */
```
Comment (jwaeab) [inline core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/utils/StatusLogger.java:46]: Clean up this comment a bit.

File: core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/utils/TestStatusLogger.java [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+
+
+        List<ILoggingEvent> logsList = listAppender.list;
+
+        long count = logsList.stream()
+                .count();
+
+        String logMessage = logsList.get(0).getFormattedMessage();
+        Level logLevel = logsList.get(0).getLevel();
+        assertEquals("Total repair success in node till now is:1", logMessage);
```
Comment (jwaeab) [inline core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/utils/TestStatusLogger.java:114]: Space after ':' missing (with the fix commented above)

File: core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/utils/TestStatusLogger.java [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+        myTableRepairMetricsImpl.lastRepairedAt(tableReference2, expectedLastRepaired2);
+        StatusLogger.log(myMeterRegistry);
+        List<ILoggingEvent> logsList = listAppender.list;
+
+        long count = logsList.stream()
+                .count();
+
+        String logMessage = logsList.get(0).getFormattedMessage();
+        Level logLevel = logsList.get(0).getLevel();
+        assertEquals("Node last repaired at",logMessage.substring(0, 21));
```
Comment (jwaeab) [inline core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/utils/TestStatusLogger.java:162]: space before logMessage.substring(...)

File: core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/utils/TestStatusLogger.java [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+        assertEquals(Level.DEBUG, logLevel);
+        assertEquals(1, count);
+    }
+
+    @Test
+    public void testLogForNodeRemainingRepairTimeStatusLogs()
+    {
+        TableReference tableReference = tableReference(TEST_KEYSPACE, TEST_TABLE1);
```
Comment (jwaeab) [inline core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/utils/TestStatusLogger.java:170]: Would name this as tableReference1, to keep the consistency (as for the rest of this test method)

File: core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/utils/TestStatusLogger.java [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+
+        String logMessage = logsList.get(0).getFormattedMessage();
+        Level logLevel = logsList.get(0).getLevel();
+        Double expectedRepairTime = (double) (expectedRemainingRepairTime+expectedRemainingRepairTime2)/1000;
+        assertEquals("Remaining time for node repair :".concat(expectedRepairTime.toString()),logMessage);
+        assertEquals(Level.DEBUG, logLevel);
+        assertEquals(1, count);
+    }
+
+
```
Comment (jwaeab) [inline core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/utils/TestStatusLogger.java:191]: Remove the extra line breaks.

File: core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/utils/TestStatusLogger.java [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+        assertEquals("Node Repair Ratio is :: 1.0", logMessage);
+        assertEquals(Level.DEBUG, logLevel);
+        assertEquals(1, count);
+
+    }
+
+@Test
+public void testLogForLastRepairedAtStatusLogs()
+    {
+        TableReference tableReference = tableReference(TEST_KEYSPACE, TEST_TABLE1);
```
Comment (jwaeab) [inline core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/utils/TestStatusLogger.java:143]: tableReference1

**Review by jwaeab — CHANGES_REQUESTED** [review]

**Comment by manmagic3** [comment]
Hi Johan, I still need to address some comments from Macan and also need to write more test cases.

**Comment by manmagic3** [comment]
Johan added the test cases and made some changes as per some comments by Makan. You can proceed with review.

File: core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/metrics/TestMetricInspector.java [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+                1, 1,5000);
+    }
+
+    @Test
+    public void testInspectMeterRegistryForRepairFailuresWhenFailureThresholdIsBroken()
+    {
+        TableReference tableReference = tableReference(TEST_KEYSPACE, TEST_TABLE1);
+        long expectedRepairTime = 12345L;
+        myTableRepairMetricsImpl.repairSession(tableReference, expectedRepairTime, TimeUnit.MILLISECONDS, false);
+        TableReference tableReference1 = tableReference(TEST_KEYSPACE, TEST_TABLE1);
```
Comment (jwaeab) [inline core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/metrics/TestMetricInspector.java:79]: You're not using "tableReference1" in this test? Why?

It is confusing when you name some variables without a number and some with and they're essentially referencing to the same table (TEST_TABLE1). Eg:
tableReference
tableReference1

Why not?
tableReference1
tableReference2

File: core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/metrics/TestMetricInspector.java [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+    }
+
+    @Test
+    public void testStartInspectionMethod() throws InterruptedException
+    {
+        assertEquals(0, myMetricInspector.getTotalRecordFailures());
+        TableReference tableReference = tableReference(TEST_KEYSPACE, TEST_TABLE1);
+        long expectedRepairTime = 12345L;
+        myTableRepairMetricsImpl.repairSession(tableReference, expectedRepairTime, TimeUnit.MILLISECONDS, false);
+        TableReference tableReference1 = tableReference(TEST_KEYSPACE, TEST_TABLE1);
```
Comment (jwaeab) [inline core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/metrics/TestMetricInspector.java:111]: Again, why tableReference1, when it is not being referenced in the test case?

File: core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/metrics/TestMetricInspector.java [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+    }
+
+    @Test
+    public void testInspectMeterRegistryForRepairFailuresWhenFailureThresholdIsBroken()
+    {
+        TableReference tableReference = tableReference(TEST_KEYSPACE, TEST_TABLE1);
+        long expectedRepairTime = 12345L;
+        myTableRepairMetricsImpl.repairSession(tableReference, expectedRepairTime, TimeUnit.MILLISECONDS, false);
+        TableReference tableReference1 = tableReference(TEST_KEYSPACE, TEST_TABLE1);
+        expectedRepairTime = 12345L;
```
Comment (jwaeab) [inline core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/metrics/TestMetricInspector.java:80]: Why setting it to the same value?

File: core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/utils/TestStatusLogger.java [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+        myTableRepairMetricsImpl.repairState(tableReference, 1, 0);
+        // Call the method to be tested
+        StatusLogger.log(myMeterRegistry);
+        List<ILoggingEvent> logsList = listAppender.list;
+        long count = logsList.stream().count();
+        String logMessage = logsList.get(0).getFormattedMessage();
+        Level logLevel = logsList.get(0).getLevel();
+        assertEquals("Node Repair Ratio is: 1.0", logMessage);
```
Comment (jwaeab) [inline core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/utils/TestStatusLogger.java:122]: Why capitol R? "Node repair ratio"

File: core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/utils/StatusLogger.java [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+        if (nodeRemainingRepairTime != null)
+        {
+            LOG.debug("Remaining time for node repair: {}", nodeRemainingRepairTime.value());
+        }
+
+        Gauge nodeRepairedRatio = myMeterRegistry.find(NODE_REPAIRED_RATIO)
+                .gauge();
+        if (nodeRepairedRatio != null)
+        {
+            LOG.debug("Node Repair Ratio is: {}", nodeRepairedRatio.value());
```
Comment (jwaeab) [inline core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/utils/StatusLogger.java:73]: Why capitol R's?

File: core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/metrics/MetricInspector.java [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+    public void stopInspection()
+    {
+        if (timer != null)
+        {
+            timer.cancel();
+        }
+    }
+
+    @VisibleForTesting
+    void inspectMeterRegistryForRepairFailures()
```
Comment (jwaeab) [inline core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/metrics/MetricInspector.java:89]: Fix code formatting for this method.

File: core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/metrics/MetricInspector.java [hunk 18]
```diff
@@ -0,0 +1,126 @@
...
+    private long myTotalRecordFailures = 0;
+    private final int myRepairFailureThreshold;
+    private final long myRepairFailureTimeWindow;
+    private final long myTriggerIntervalForMetricInspection;
+    private LocalDateTime myRecordingStartTimestamp = LocalDateTime.now();
+    private Timer timer;
+
+    @VisibleForTesting
+    long getRepairFailuresCountSinceLastReport()
+     {
```
Comment (jwaeab) [inline core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/metrics/MetricInspector.java:40]: Formatting

File: core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/utils/TestStatusLogger.java [hunk 19]
```diff
@@ -0,0 +1,168 @@
...
+        long expectedRemainingRepairTime = 10L;
+        long expectedRemainingRepairTime2 = 20L;
+        myTableRepairMetricsImpl.remainingRepairTime(tableReference1, expectedRemainingRepairTime);
+        myTableRepairMetricsImpl.remainingRepairTime(tableReference2, expectedRemainingRepairTime2);
+        StatusLogger.log(myMeterRegistry);
+        List<ILoggingEvent> logsList = listAppender.list;
+        long count = logsList.stream().count();
+        String logMessage = logsList.get(0).getFormattedMessage();
+        Level logLevel = logsList.get(0).getLevel();
+        Double expectedRepairTime = (double) (expectedRemainingRepairTime+expectedRemainingRepairTime2) / 1000;
```
Comment (jwaeab) [inline core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/utils/TestStatusLogger.java:163]: Formatting "...RepairTime1 + expectedRemaining..."

File: core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/metrics/TestMetricInspector.java [hunk 20]
```diff
@@ -0,0 +1,125 @@
...
+        // Use composite registry here to simulate real world scenario where we have multiple registries
+        CompositeMeterRegistry compositeMeterRegistry = new CompositeMeterRegistry();
+        // Need at least one registry present in composite to record metrics
+        compositeMeterRegistry.add(new SimpleMeterRegistry());
+        myMeterRegistry = compositeMeterRegistry;
+        myTableRepairMetricsImpl = TableRepairMetricsImpl.builder()
+                .withTableStorageStates(myTableStorageStates)
+                .withMeterRegistry(myMeterRegistry)
+                .build();
+        myMetricInspector = new MetricInspector(myMeterRegistry,
```
Comment (jwaeab) [inline core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/metrics/TestMetricInspector.java:69]: Code formatting

File: core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/metrics/TestMetricInspector.java [hunk 21]
```diff
@@ -0,0 +1,125 @@
...
+import static org.junit.Assert.assertEquals;
+
+
+@RunWith(MockitoJUnitRunner.class)
+public class TestMetricInspector
+{
+    private static final String TEST_KEYSPACE = "test_keyspace";
+    private static final String TEST_TABLE1 = "test_table1";
+    private static final String TEST_TABLE2 = "test_table2";
+    @Mock
```
Comment (jwaeab) [inline core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/metrics/TestMetricInspector.java:45]: Perhaps a blank line here

File: core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/metrics/MetricInspector.java [hunk 22]
```diff
@@ -0,0 +1,126 @@
...
+        resetRepairFailureCount();
+    }
+
+    /*
+     * If in defined time window, number of repair failure has not crossed the configured number, then
+     * reset failure count for new timed window. Reinitialize time window.
+     */
+
+    @VisibleForTesting
+    void resetRepairFailureCount()
```
Comment (jwaeab) [inline core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/metrics/MetricInspector.java:115]: Formatting

File: application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/ECChronosInternals.java [hunk 23]
```diff
@@ -108,11 +111,18 @@ public ECChronosInternals(final Config configuration,
                     .withTableStorageStates(myTableStorageStatesImpl)
                     .withMeterRegistry(meterRegistry)
                     .build();
+
+            myMetricInspector = new MetricInspector(meterRegistry,
+                    configuration.getStatisticsConfig().getRepairFailuresCount(),
+                    configuration.getStatisticsConfig().getRepairFailuresTimeWindowInMinutes(),
```
Comment (jwaeab) [inline application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/ECChronosInternals.java:117]: Why do you need this specialized method here?

File: application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/ECChronosInternals.java [hunk 24]
```diff
@@ -108,11 +111,18 @@ public ECChronosInternals(final Config configuration,
...
+                    configuration.getStatisticsConfig().getTriggerIntervalForMetricInspectionInMillSeconds());
```
Comment (jwaeab) [inline application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/ECChronosInternals.java:118]: Why do you need this specialized method here?

**Comment by manmagic3** [comment]
Documentation updatated, comments addressed.

File: application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/config/metrics/StatisticsConfig.java [hunk 25]
```diff
@@ -14,24 +14,32 @@
...
 import com.fasterxml.jackson.annotation.JsonProperty;
 
 import java.io.File;
+import java.util.concurrent.TimeUnit;
 
 public class StatisticsConfig
 {
+    private static final int DEFAULT_FAILURE_TIME_WINDOW_IN_MINUTES = 30;
+    private static final int DEFAULT_TRIGGER_INTERVAL_FOR_METRIC_INSPECTION_IN_SECONDS = 5;
+    private static final int DEFAULT_REPAIR_FAILURE_COUNT = 5;
```
Comment (jwaeab) [inline application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/config/metrics/StatisticsConfig.java:27]: "failures", to keep it consistent; DEFAULT_REPAIR_FAILURES_COUNT

File: application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/config/metrics/StatisticsConfig.java [hunk 26]
```diff
@@ -14,24 +14,32 @@
...
 
+import com.ericsson.bss.cassandra.ecchronos.application.config.Interval;
 import com.fasterxml.jackson.annotation.JsonProperty;
 
 import java.io.File;
+import java.util.concurrent.TimeUnit;
 
 public class StatisticsConfig
 {
+    private static final int DEFAULT_FAILURE_TIME_WINDOW_IN_MINUTES = 30;
```
Comment (jwaeab) [inline application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/config/metrics/StatisticsConfig.java:25]: "failures", to keep it consistent; DEFAULT_FAILURES_TIME_WINDOW_IN_MINUTES

File: core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/metrics/MetricInspector.java [hunk 27]
```diff
@@ -0,0 +1,124 @@
...
+        return myTotalRecordFailures;
+    }
+
+    @VisibleForTesting
+    LocalDateTime getRecordingStartTimestamp()
+    {
+        return myRecordingStartTimestamp;
+    }
+
+       public MetricInspector(final MeterRegistry meterRegistry,
```
Comment (jwaeab) [inline core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/metrics/MetricInspector.java:55]: Formatting

File: core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/metrics/MetricInspector.java [hunk 28]
```diff
@@ -0,0 +1,124 @@
...
+        if (timer != null)
+        {
+            timer.cancel();
+        }
+    }
+
+    @VisibleForTesting
+    void inspectMeterRegistryForRepairFailures()
+    {
+        io.micrometer.core.instrument.Timer nodeRepairSessions = myMeterRegistry
```
Comment (jwaeab) [inline core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/metrics/MetricInspector.java:90]: Formatting in the entire method should be cleaned up.

File: docs/METRICS.md [hunk 29]
```diff
@@ -742,4 +742,15 @@ repair_sessions_seconds_max{keyspace="ks1",successful="true",table="tbl1",} 0.0
 repair_sessions_seconds_count{keyspace="ks1",successful="false",table="tbl1",} 793.0
 repair_sessions_seconds_sum{keyspace="ks1",successful="false",table="tbl1",} 5.317104685
 repair_sessions_seconds_max{keyspace="ks1",successful="false",table="tbl1",} 0.0
-```
\ No newline at end of file
+```
+
+## Metric Status Logger
+Whenever metric is enabled, a logger is triggered which monitors metrics for
```
Comment (jwaeab) [inline docs/METRICS.md:748]: monitor

File: docs/METRICS.md [hunk 30]
```diff
@@ -742,4 +742,15 @@ repair_sessions_seconds_max{keyspace="ks1",successful="true",table="tbl1",} 0.0
...
+repair failures within a defined time window. If number of repair failures 
+overshoot the number(`repair_failures_count`) configured in ecc.yml within a time
```
Comment (jwaeab) [inline docs/METRICS.md:750]: ... within the time ...

**Review by jwaeab — CHANGES_REQUESTED** [review]

**Comment by manmagic3** [comment]
Hi Johan I have closed all review comments.

**Review by jwaeab — APPROVED** [review]

**Review by jwaeab — CHANGES_REQUESTED** [review]
You need to update CHANGES.md as well as have a look at the conflicts

**Review by jwaeab — DISMISSED** [review]

**Review by tommystendahl — APPROVED** [review]

**Pull request merged**